src/binary/merkle_tree.rs

```diff
@@ -107,11 +106,21 @@ where
         side_positions.reverse(); // Reorder side positions from leaf to root.
         side_positions.pop(); // The last side position is the root; remove it.
 
+        // Allocate scratch storage to store temporary nodes when building the
+        // root.
+        let mut scratch_storage = StorageMap::<NodesTable>::new();
+        let root_node = self.root_node(&mut scratch_storage)?.unwrap();
+
+        // Get side nodes. First, we check the scratch storage. If the side node
+        // is not found in scratch storage, we then check main storage. Finally,
+        // if the side node is not found in main storage, we exit with a load
+        // error.
         for side_position in side_positions {
             let key = side_position.in_order_index();
-            let primitive = self
-                .storage
-                .get(&key)?
+            let primitive = scratch_storage
+                .get(&key)
+                .unwrap()
```

Comment:
```suggestion
                .get(&key)?
```

Reply:
It's an option value so the `?` operator cannot be used

Reply:
It is not an `Option`, it is `StorageResult`=)

Reply:
Originally, this was a `Result<Option<T>, core::convert::Infallible>`. In either case, I could not use `?` because you cannot convert `Infallible` to the desired error type, nor can you ignore it. So I introduced an infallible method signature that simply returns the `Option<T>`. this way, no `unwrap`, `expect`, or `?` is needed.